File: frameworks/php-laravel/tests/Feature/AdvancedIntegrationTest.php
<?php

namespace Tests\Feature;

use Tests\TestCase;
use Tests\Traits\DatabaseTrait;
use App\Models\User;
use App\Models\Post;
use App\Models\Comment;

/**
 * AdvancedIntegrationTest - Advanced flows, relationships, and data consistency
 *
 * Additional tests to boost Laravel suite to 75+ tests
 */
class AdvancedIntegrationTest extends TestCase
{
    use DatabaseTrait;

    // ========================================================================
    // Multi-User Relationship Tests
    // ========================================================================

    public function test_multiple_authors_have_separate_posts(): void
    {
        // Arrange
        $author1 = $this->createTestUser("author1", "author1@example.com", "Author 1");
        $author2 = $this->createTestUser("author2", "author2@example.com", "Author 2");
        $author3 = $this->createTestUser("author3", "author3@example.com", "Author 3");

        $post1 = $this->createTestPost($author1->pk_user, "Author 1 Post", "Content 1");
        $post2 = $this->createTestPost($author2->pk_user, "Author 2 Post", "Content 2");
        $post3 = $this->createTestPost($author3->pk_user, "Author 3 Post", "Content 3");

        // Act
        $response1 = $this->getJson("/api/posts/by-author/{$author1->id}");
        $response2 = $this->getJson("/api/posts/by-author/{$author2->id}");
        $response3 = $this->getJson("/api/posts/by-author/{$author3->id}");

        // Assert
        $this->assertEquals(1, count($response1->json()));
        $this->assertEquals(1, count($response2->json()));
        $this->assertEquals(1, count($response3->json()));
    }

    public function test_post_relationships_are_independent(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com", "Author");
        $post1 = $this->createTestPost($author->pk_user, "Post 1", "Content 1");
        $post2 = $this->createTestPost($author->pk_user, "Post 2", "Content 2");

        // Act
        $response1 = $this->getJson("/api/posts/{$post1->id}");
        $response2 = $this->getJson("/api/posts/{$post2->id}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();
        $this->assertNotEquals($data1['id'], $data2['id']);
        $this->assertEquals("Post 1", $data1['title']);
        $this->assertEquals("Post 2", $data2['title']);
    }

    // ========================================================================
    // Pagination Edge Cases
    // ========================================================================

    public function test_pagination_boundary_at_exact_limit(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        for ($i = 0; $i < 20; $i++) {
            $this->createTestPost($author->pk_user, "Post $i", "Content $i");
        }

        // Act
        $response = $this->getJson("/api/posts/by-author/{$author->id}?page=0&size=20");

        // Assert
        $this->assertEquals(20, count($response->json()));
    }

    public function test_pagination_page_alignment(): void
    {
        // Arrange
        for ($i = 0; $i < 30; $i++) {
            $user = $this->createTestUser("user$i", "user$i@example.com");
            if ($i % 5 == 0) {
                $this->createTestPost($user->pk_user, "Post", "Content");
            }
        }

        // Act
        $page1 = $this->getJson("/api/users?page=0&size=15");
        $page2 = $this->getJson("/api/users?page=1&size=15");

        // Assert
        $this->assertEquals(15, count($page1->json()));
        $this->assertEquals(15, count($page2->json()));
    }

    // ========================================================================
    // Field Immutability Tests
    // ========================================================================

    public function test_username_remains_immutable(): void
    {
        // Arrange
        $user = $this->createTestUser("alice", "alice@example.com", "Alice");
        $originalUsername = $user->username;

        // Act
        $user->update(['bio' => 'New bio']);

        // Verify
        $response = $this->getJson("/api/users/{$user->id}");

        // Assert
        $data = $response->json();
        $this->assertEquals($originalUsername, $data['username']);
    }

    public function test_id_remains_immutable(): void
    {
        // Arrange
        $user = $this->createTestUser("alice", "alice@example.com", "Alice");
        $originalId = $user->id;

        // Act
        $user->update(['bio' => 'New bio']);

        // Verify
        $response = $this->getJson("/api/users/{$user->id}");

        // Assert
        $data = $response->json();
        $this->assertEquals($originalId, $data['id']);
    }

    // ========================================================================
    // Data Type Validation
    // ========================================================================

    public function test_all_user_ids_are_uuid(): void
    {
        // Arrange
        $users = [
            $this->createTestUser("alice", "alice@example.com"),
            $this->createTestUser("bob", "bob@example.com"),
            $this->createTestUser("charlie", "charlie@example.com"),
        ];

        // Act & Assert
        foreach ($users as $user) {
            $parts = explode('-', $user->id);
            $this->assertCount(5, $parts);
        }
    }

    public function test_all_post_ids_are_uuid(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        $posts = [
            $this->createTestPost($author->pk_user, "Post 1", "Content 1"),
            $this->createTestPost($author->pk_user, "Post 2", "Content 2"),
            $this->createTestPost($author->pk_user, "Post 3", "Content 3"),
        ];

        // Act & Assert
        foreach ($posts as $post) {
            $parts = explode('-', $post->id);
            $this->assertCount(5, $parts);
        }
    }

    // ========================================================================
    // Response Structure Consistency
    // ========================================================================

    public function test_list_and_detail_structure_match(): void
    {
        // Arrange
        $user = $this->createTestUser("alice", "alice@example.com", "Alice", "Alice bio");

        // Act
        $listResponse = $this->getJson("/api/users");
        $detailResponse = $this->getJson("/api/users/{$user->id}");

        // Assert
        $listData = $listResponse->json();
        $detailData = $detailResponse->json();

        $listUser = collect($listData)->firstWhere('id', $user->id);

        $this->assertEquals($listUser['username'], $detailData['username']);
        $this->assertEquals($listUser['fullName'], $detailData['fullName']);
        $this->assertEquals($listUser['bio'] ?? null, $detailData['bio']);
    }

    public function test_post_list_and_detail_structure_match(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        $post = $this->createTestPost($author->pk_user, "Test Post", "Test Content");

        // Act
        $listResponse = $this->getJson("/api/posts");
        $detailResponse = $this->getJson("/api/posts/{$post->id}");

        // Assert
        $listData = $listResponse->json();
        $detailData = $detailResponse->json();

        $listPost = collect($listData)->firstWhere('id', $post->id);

        $this->assertEquals($listPost['title'], $detailData['title']);
        $this->assertEquals($listPost['content'], $detailData['content']);
    }

    // ========================================================================
    // Timestamp Consistency Tests
    // ========================================================================

    public function test_user_timestamps_are_set(): void
    {
        // Arrange
        $user = $this->createTestUser("alice", "alice@example.com", "Alice");

        // Act & Assert
        $this->assertNotNull($user->created_at);
        $this->assertIsObject($user->created_at);
    }

    public function test_post_timestamps_are_set(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        $post = $this->createTestPost($author->pk_user, "Test Post", "Content");

        // Act & Assert
        $this->assertNotNull($post->created_at);
        $this->assertIsObject($post->created_at);
    }

    public function test_timestamps_returned_in_responses(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        $post = $this->createTestPost($author->pk_user, "Test Post", "Content");

        // Act
        $response = $this->getJson("/api/posts/{$post->id}");

        // Assert
        $data = $response->json();
        $this->assertNotNull($data['createdAt']);
    }

    // ========================================================================
    // Null Field Consistency Tests
    // ========================================================================

    public function test_null_bio_consistency(): void
    {
        // Arrange
        $user1 = $this->createTestUser("alice", "alice@example.com", "Alice", null);
        $user2 = $this->createTestUser("bob", "bob@example.com", "Bob", "Bob's bio");

        // Act
        $response1 = $this->getJson("/api/users/{$user1->id}");
        $response2 = $this->getJson("/api/users/{$user2->id}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();

        $this->assertNull($data1['bio']);
        $this->assertNotNull($data2['bio']);
    }

    public function test_empty_string_vs_null_bio(): void
    {
        // Arrange
        $user1 = $this->createTestUser("alice", "alice@example.com", "Alice", "");
        $user2 = $this->createTestUser("bob", "bob@example.com", "Bob", null);

        // Act
        $response1 = $this->getJson("/api/users/{$user1->id}");
        $response2 = $this->getJson("/api/users/{$user2->id}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();

        $this->assertEquals("", $data1['bio']);
        $this->assertNull($data2['bio']);
    }

    // ========================================================================
    // Query Result Consistency Tests
    // ========================================================================

    public function test_same_user_returns_same_data(): void
    {
        // Arrange
        $user = $this->createTestUser("alice", "alice@example.com", "Alice", "Alice bio");
        $userId = $user->id;

        // Act
        $response1 = $this->getJson("/api/users/{$userId}");
        $response2 = $this->getJson("/api/users/{$userId}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();

        $this->assertEquals($data1['id'], $data2['id']);
        $this->assertEquals($data1['username'], $data2['username']);
        $this->assertEquals($data1['bio'], $data2['bio']);
    }

    public function test_same_post_returns_same_data(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com");
        $post = $this->createTestPost($author->pk_user, "Test Post", "Test Content");
        $postId = $post->id;

        // Act
        $response1 = $this->getJson("/api/posts/{$postId}");
        $response2 = $this->getJson("/api/posts/{$postId}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();

        $this->assertEquals($data1['id'], $data2['id']);
        $this->assertEquals($data1['title'], $data2['title']);
        $this->assertEquals($data1['content'], $data2['content']);
    }

    // ========================================================================
    // Trinity Pattern Consistency
    // ========================================================================

    public function test_post_author_id_references_user(): void
    {
        // Arrange
        $author = $this->createTestUser("author", "author@example.com", "Author");
        $post = $this->createTestPost($author->pk_user, "Test Post", "Content");

        // Act
        $response = $this->getJson("/api/posts/{$post->id}");

        // Assert
        $data = $response->json();
        $this->assertEquals($author->id, $data['authorId']);
    }

    public function test_multiple_posts_reference_correct_authors(): void
    {
        // Arrange
        $author1 = $this->createTestUser("author1", "author1@example.com");
        $author2 = $this->createTestUser("author2", "author2@example.com");

        $post1 = $this->createTestPost($author1->pk_user, "Post 1", "Content");
        $post2 = $this->createTestPost($author2->pk_user, "Post 2", "Content");

        // Act
        $response1 = $this->getJson("/api/posts/{$post1->id}");
        $response2 = $this->getJson("/api/posts/{$post2->id}");

        // Assert
        $data1 = $response1->json();
        $data2 = $response2->json();

        $this->assertEquals($author1->id, $data1['authorId']);
        $this->assertEquals($author2->id, $data2['authorId']);
    }
}

File: frameworks/ruby-rails/test/controllers/advanced_integration_test.rb
require "test_helper"

class AdvancedIntegrationTest < ActionDispatch::IntegrationTest
  # ========================================================================
  # Multi-User Relationship Tests
  # ========================================================================

  test "multiple authors have separate posts" do
    # Arrange
    author1 = create_test_user("author1", "author1@example.com", "Author 1")
    author2 = create_test_user("author2", "author2@example.com", "Author 2")
    author3 = create_test_user("author3", "author3@example.com", "Author 3")

    post1 = create_test_post(author1.pk_user, "Author 1 Post", "Content 1")
    post2 = create_test_post(author2.pk_user, "Author 2 Post", "Content 2")
    post3 = create_test_post(author3.pk_user, "Author 3 Post", "Content 3")

    # Act
    response1 = get "/api/posts/by-author/#{author1.id}"
    response2 = get "/api/posts/by-author/#{author2.id}"
    response3 = get "/api/posts/by-author/#{author3.id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)
    data3 = JSON.parse(response3.body)

    assert_equal 1, data1.length
    assert_equal 1, data2.length
    assert_equal 1, data3.length
  end

  test "post relationships are independent" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post1 = create_test_post(author.pk_user, "Post 1", "Content 1")
    post2 = create_test_post(author.pk_user, "Post 2", "Content 2")

    # Act
    response1 = get "/api/posts/#{post1.id}"
    response2 = get "/api/posts/#{post2.id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_not_equal data1["id"], data2["id"]
    assert_equal "Post 1", data1["title"]
    assert_equal "Post 2", data2["title"]
  end

  # ========================================================================
  # Pagination Edge Cases
  # ========================================================================

  test "pagination boundary at exact limit" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    15.times { |i| create_test_post(author.pk_user, "Post #{i}", "Content #{i}") }

    # Act
    response = get "/api/posts/by-author/#{author.id}?page=0&size=15"

    # Assert
    data = JSON.parse(response.body)
    assert_equal 15, data.length
  end

  test "pagination page alignment" do
    # Arrange
    30.times { |i| create_test_user("user#{i}", "user#{i}@example.com") }

    # Act
    response1 = get "/api/users?page=0&size=15"
    response2 = get "/api/users?page=1&size=15"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_equal 15, data1.length
    assert_equal 15, data2.length
  end

  # ========================================================================
  # Field Immutability Tests
  # ========================================================================

  test "username remains immutable" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    original_username = user.username

    # Act
    user.update(bio: "New bio")

    # Verify
    response = get "/api/users/#{user.id}"

    # Assert
    data = JSON.parse(response.body)
    assert_equal original_username, data["username"]
  end

  test "id remains immutable" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")
    original_id = user.id

    # Act
    user.update(bio: "New bio")

    # Verify
    response = get "/api/users/#{user.id}"

    # Assert
    data = JSON.parse(response.body)
    assert_equal original_id, data["id"]
  end

  # ========================================================================
  # Data Type Validation
  # ========================================================================

  test "all user ids are uuid" do
    # Arrange
    users = [
      create_test_user("alice", "alice@example.com"),
      create_test_user("bob", "bob@example.com"),
      create_test_user("charlie", "charlie@example.com"),
    ]

    # Act & Assert
    users.each do |user|
      parts = user.id.split("-")
      assert_equal 5, parts.length
    end
  end

  test "all post ids are uuid" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    posts = [
      create_test_post(author.pk_user, "Post 1", "Content 1"),
      create_test_post(author.pk_user, "Post 2", "Content 2"),
      create_test_post(author.pk_user, "Post 3", "Content 3"),
    ]

    # Act & Assert
    posts.each do |post|
      parts = post.id.split("-")
      assert_equal 5, parts.length
    end
  end

  # ========================================================================
  # Response Structure Consistency
  # ========================================================================

  test "list and detail structure match" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice", "Alice bio")

    # Act
    list_response = get "/api/users"
    detail_response = get "/api/users/#{user.id}"

    # Assert
    list_data = JSON.parse(list_response.body)
    detail_data = JSON.parse(detail_response.body)

    list_user = list_data.find { |u| u["id"] == user.id }

    assert_equal list_user["username"], detail_data["username"]
    assert_equal list_user["fullName"], detail_data["fullName"]
  end

  test "post list and detail structure match" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    post = create_test_post(author.pk_user, "Test Post", "Test Content")

    # Act
    list_response = get "/api/posts"
    detail_response = get "/api/posts/#{post.id}"

    # Assert
    list_data = JSON.parse(list_response.body)
    detail_data = JSON.parse(detail_response.body)

    list_post = list_data.find { |p| p["id"] == post.id }

    assert_equal list_post["title"], detail_data["title"]
    assert_equal list_post["content"], detail_data["content"]
  end

  # ========================================================================
  # Timestamp Consistency Tests
  # ========================================================================

  test "user timestamps are set" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice")

    # Act & Assert
    assert_not_nil user.created_at
  end

  test "post timestamps are set" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    post = create_test_post(author.pk_user, "Test Post", "Content")

    # Act & Assert
    assert_not_nil post.created_at
  end

  test "timestamps returned in responses" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    post = create_test_post(author.pk_user, "Test Post", "Content")

    # Act
    response = get "/api/posts/#{post.id}"

    # Assert
    data = JSON.parse(response.body)
    assert_not_nil data["createdAt"]
  end

  # ========================================================================
  # Null Field Consistency Tests
  # ========================================================================

  test "null bio consistency" do
    # Arrange
    user1 = create_test_user("alice", "alice@example.com", "Alice", nil)
    user2 = create_test_user("bob", "bob@example.com", "Bob", "Bob's bio")

    # Act
    response1 = get "/api/users/#{user1.id}"
    response2 = get "/api/users/#{user2.id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_nil data1["bio"]
    assert_not_nil data2["bio"]
  end

  test "empty string vs null bio" do
    # Arrange
    user1 = create_test_user("alice", "alice@example.com", "Alice", "")
    user2 = create_test_user("bob", "bob@example.com", "Bob", nil)

    # Act
    response1 = get "/api/users/#{user1.id}"
    response2 = get "/api/users/#{user2.id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_equal "", data1["bio"]
    assert_nil data2["bio"]
  end

  # ========================================================================
  # Query Result Consistency Tests
  # ========================================================================

  test "same user returns same data" do
    # Arrange
    user = create_test_user("alice", "alice@example.com", "Alice", "Alice bio")
    user_id = user.id

    # Act
    response1 = get "/api/users/#{user_id}"
    response2 = get "/api/users/#{user_id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_equal data1["id"], data2["id"]
    assert_equal data1["username"], data2["username"]
    assert_equal data1["bio"], data2["bio"]
  end

  test "same post returns same data" do
    # Arrange
    author = create_test_user("author", "author@example.com")
    post = create_test_post(author.pk_user, "Test Post", "Test Content")
    post_id = post.id

    # Act
    response1 = get "/api/posts/#{post_id}"
    response2 = get "/api/posts/#{post_id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_equal data1["id"], data2["id"]
    assert_equal data1["title"], data2["title"]
    assert_equal data1["content"], data2["content"]
  end

  # ========================================================================
  # Trinity Pattern Consistency
  # ========================================================================

  test "post author id references user" do
    # Arrange
    author = create_test_user("author", "author@example.com", "Author")
    post = create_test_post(author.pk_user, "Test Post", "Content")

    # Act
    response = get "/api/posts/#{post.id}"

    # Assert
    data = JSON.parse(response.body)
    assert_equal author.id, data["authorId"]
  end

  test "multiple posts reference correct authors" do
    # Arrange
    author1 = create_test_user("author1", "author1@example.com")
    author2 = create_test_user("author2", "author2@example.com")

    post1 = create_test_post(author1.pk_user, "Post 1", "Content")
    post2 = create_test_post(author2.pk_user, "Post 2", "Content")

    # Act
    response1 = get "/api/posts/#{post1.id}"
    response2 = get "/api/posts/#{post2.id}"

    # Assert
    data1 = JSON.parse(response1.body)
    data2 = JSON.parse(response2.body)

    assert_equal author1.id, data1["authorId"]
    assert_equal author2.id, data2["authorId"]
  end
end
